plugins/Lv2Effect/Lv2Effect.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * Lv2Effect.cpp - implementation of LV2 effect
+ *
+ * Copyright (c) 2018-2019 Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Lv2Effect.h"
+
+#include <QDebug>
+#include <lv2.h>
+
+#include "Lv2SubPluginFeatures.h"
+
+#include "embed.h"
+#include "plugin_export.h"
+
+
+
+
+Plugin::Descriptor PLUGIN_EXPORT lv2effect_plugin_descriptor =
+{
+	STRINGIFY(PLUGIN_NAME),
+	"LV2",
+	QT_TRANSLATE_NOOP("pluginBrowser",
+		"plugin for using arbitrary LV2-effects inside LMMS."),
+	"Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	nullptr,
+	new Lv2SubPluginFeatures(Plugin::Effect)
+};
+
+
+
+
+Lv2Effect::Lv2Effect(Model* parent, const Descriptor::SubPluginFeatures::Key *key) :
+	Effect(&lv2effect_plugin_descriptor, parent, key),
+	m_controls(this, key->attributes["uri"])
+{
+}
+
+
+
+
+Lv2Effect::~Lv2Effect()
+{
+}
+
+
+
+
+bool Lv2Effect::processAudioBuffer(sampleFrame *buf, const fpp_t frames)
+{
+	if (!isEnabled() || !isRunning()) { return false; }
+
+	Lv2FxControls& ctrl = m_controls;
+
+	ctrl.copyBuffersFromLmms(buf, frames);
+	m_controls.copyModelsFromLmms();
+
+//	m_pluginMutex.lock();
+	ctrl.run(frames);
+//	m_pluginMutex.unlock();
+
+	ctrl.copyBuffersToLmms(buf, frames);
+
+	double outSum = .0;
+	for(fpp_t f = 0; f < frames; ++f)
+	{
+		double l = static_cast<double>(buf[f][0]);
+		double r = static_cast<double>(buf[f][1]);
+		outSum += l*l + r*r;
+	}
+	checkGate(outSum);
```

Comment:
Makes sense, though I wonder why we don't take the `sqrt` after division. Probably because `checkGate` calculates in squares? At least, it does calculate `_out_sum - gate()`, and `gate()` seems to be squared. So I assume `checkGate` expects the squared RMS value over the buffer?

Strangely, all but one effect divide by `frames`: `CrossoverEQ.cpp`. Maybe this is buggy, too?

Reply:
> So I assume `checkGate` expects the squared RMS value over the buffer?

It looks like that to me too; perhaps this should go in a comment somewhere?

> Strangely, all but one effect divide by `frames`: `CrossoverEQ.cpp`. Maybe this is buggy, too?

It seems to calculate `outSum` incorrectly too - it's assigned rather than incremented in the loop, so only the last sample is actually checked.

Reply:
OK, I fixed it in Lv2 and moved the rest to #5460 and #5461 , so we can resolve the comment here.

Reply:
What I just see when comparing to other effects, `Lv2Effect` does not yet take wet/dry-levels into account. Should I still implement this?

Reply:
You mentioned in [this comment](https://github.com/LMMS/lmms/pull/4899/files#r277143526) that you wanted to integrate this with the LV2 parameters extension in a later PR. It might be a good idea to implement it the normal way in the meantime though.

Reply:
Oh good you recalled.

I can implement that it now works for all effects "normally". Later for only those that offer `lv2:dryLevel`, we could connect that parameter to the `Effect::dryLevel` somehow and for the rest do the "normal" computation. Do you mean that?

It's easy enough to do for now. I just hope users won't be unsatisfied if the effect of the D/W knob will change when we enable `lv2:dryLevel`. Do you think this could be a problem?

Reply:
I think it should have the same effect no matter what plugin it's used with. It should be safe to connect it to an `lv2:wetDryRatio` control, but if connected to an `lv2:dryLevel` control, it would be necessary to also set a corresponding `lv2:wetLevel` control to an appropriate value. If we use one knob to control two controls, would the loss of flexibility be a problem to some users?

Also, in the future, I think it would be nice to refactor all the gate and wet/dry code so instead of each plugin implementing it individually, it's all handled in one place in the core. This is less of a concern though as it's just about the code, rather than anything that affects the user.

Another thing to consider is #4635 - this should be easy enough to support in the case of two mono processors, since the wet/dry can be set for each individually, but in the case of a single stereo processor, it would need to be handled manually.

At any rate, I think implementing it normally for now is the best way to go, and we can discuss later the advantages and disadvantages of connecting it to LV2 parameters. There's no reason for this PR to be held up by something that's not meant to go in it.

Reply:
Yeah, probably the behavior will not change when we move to the lv2 functionality for dry-wet.

> At any rate, I think implementing it normally for now is the best way to go, and we can discuss later the advantages and disadvantages of connecting it to LV2 parameters. There's no reason for this PR to be held up by something that's not meant to go in it.

Agreed.

One implementation question: Would you rather take the d/w ratio via `m_controls.copyBuffersToLmms(buf, frames)` over to `Audio::copyBuffersToCore` and do the d/w computation there, or do the d/w computation directly in `Lv2Effect::processAudioBuffer`? The latter requires a temporary buffer and and additionally `memcpy()`, which is probably still very fast. The former might be more difficult to read, not sure though.

Reply:
I agree that doing the D/W computations in `Lv2Effect::processAudioBuffer` is probably easier to read. It's more amenable to later refactoring too - e.g. implementing #4635 for LV2 effects, or moving D/W handling into the core. I think it's the best approach, unless profiling shows it to be problematically slow.